Update workspace documentation

laurapacilio opened the pull request:
## What
Update the workspaces documentation in the Language and CLI documentation sets. Specifically this PR makes the following changes:
- Pulls out the majority of the "How workspaces work" content from the language docs and puts it all together in the CLI documentation. I have added links on the language page to the CLI docs where users can find more information. 
- Removes version information about very old Terraform versions from the language documentation. We have versioned docs now, so this seems unnecessary. We're also on Terraform v1.3, which is very far removed from v0.9 that the note is referring to.
- Adds headers to help break up long blocks of content and help users better find useful information. 
- Clarifies when users should and should not use workspaces to manage multiple groups of infrastructure. Specifically, it appears that this is not appropriate when multiple sets of infrastructure need strong separation - access credentials, etc. Because workspaces within a working directory use the same backend, this type of separation is not possible with workspaces alone.
- Tries to make the content as short and concise as possible. Also revises it to adhere to style conventions.

## Why

Users are understandably confused by the fact that we currently talk about how you can use workspaces for deployments in multiple environments but then explicitly say that this is not a recommended pattern for some deployments. It's pretty hard to understand right now what's recommended and when practitioners should be using workspaces for multiple sets of infrastructure. 

## Screenshots

### Language Workspaces Page
<img width="513" alt="Screen Shot 2022-09-22 at 7 08 55 PM" src="https://user-images.githubusercontent.com/83350965/191866820-333fb2c4-f7e8-4ade-b28b-f683d5b0a170.png">

### CLI Workspaces Page

<img width="488" alt="Screen Shot 2022-09-23 at 4 32 57 PM" src="https://user-images.githubusercontent.com/83350965/192052200-1822ffc2-4db6-47bb-b086-4fb9ee76f437.png">

----

apparentlymart commented on website/docs/cli/workspaces/index.mdx:
```diff
@@ -7,72 +7,99 @@ description: >-
+
+Terraform relies on state to associate resources with real-world objects. When you run the same configuration multiple times with separate state data, Terraform can manage multiple sets of non-overlapping resources.
+
+Workspaces let you create many instances of the same infrastructure or inject different [variable values](/language/values/variables) for each instance. For example, you may want to quickly create infrastructure to test changes without affecting the production deployment.
```
I think the "let you ... inject different variable values" statement here is a bit tricky...

It reads to me like it's saying that different variable values per workspace are a built-in feature of workspaces, but that isn't true. Instead, I think it's perhaps more true to say something like this:

> Workspaces let you create many instances of the same infrastructure. You can describe small differences between those instances by setting different [variable values](/language/values/variables) for each instance.

I'm not sure this is totally clear either, so I will say the same thing in more words just in case: _If_ you choose to use multiple workspaces then you will typically _also_ need to manually specify different input variables for each distinct workspace, because otherwise they'll all be describing exactly the same infrastructure and will inevitably collide with one another. So injecting different input variables is something you start needing to do as a consequence of using workspaces, not a _benefit_ of workspaces. Hopefully you can see what I mean with all of this incoherent jibber-jabber. :upside_down_face:

laurapacilio replied:
ahhh okay this is very helpful! I'm going to push up a fix for this to make it clearer - thank you!

apparentlymart commented on website/docs/cli/workspaces/index.mdx:
```diff
@@ -7,72 +7,99 @@ description: >-
-Workspaces allow you to use the same working copy of your configuration and the
-same plugin and module caches, while still keeping separate states for each
-collection of resources you manage.
+Workspaces in the Terraform CLI refer to separate instances of [state data](/language/state) inside the same working directory.
```
I assume by including "in the Terraform CLI" here you were implicitly saying we're not talking about workspaces in Terraform Cloud, but I would recommend being more explicit about that because it's a common source of confusion.

The common way folks talk about it in the community is to assert that Terraform CLI workspaces and Terraform Cloud workspaces are _entirely separate_. That isn't strictly true -- Terraform Cloud workspaces are more like an _extension_ to Terraform CLI workspaces with a number of additional features and better usability -- but since we are intentionally not talking about the Terraform Cloud-unique features here I think it would be helpful to link right up top to some more relevant page to read if you are a Terraform Cloud user, such as [the Cloud Workspaces overview page](https://www.terraform.io/cloud-docs/workspaces).

laurapacilio replied:
This is a good point --> I can push up a fix for this!

apparentlymart commented on website/docs/cli/workspaces/index.mdx:
```diff
@@ -7,72 +7,99 @@ description: >-
+
+You can create multiple [working directories](/cli/init) to maintain multiple instances of a configuration with completely separate state data. However, Terraform installs a separate cache of plugins and modules for each working directory, so maintaining multiple directories can waste bandwidth and disk space. This approach also requires extra tasks like updating configuration from version contro ...
+
+A common use for multiple workspaces is to create a parallel, distinct copy of
+a set of infrastructure to test a set of changes before modifying production infrastructure.
+
+Non-default workspaces are often related to feature branches in version control.
+The default workspace might correspond to the `main` or `trunk` branch,
+which describes the intended state of production infrastructure. When a
+feature branch is created to develop a change, the developer of that feature
+might create a corresponding workspace and deploy into it a temporary copy
+of the main infrastructure so that they can test changes without affecting
+the production infrastructure. Once the change is merged and deployed to the
+default workspace, they destroy the test infrastructure and delete the temporary workspace.
```
These paragraphs are still hard-wrapped like the old content on this page. Did you intend to remove all of the hard-wrapping from the paragraphs here? (I've noticed you've been standardizing on not hard-wrapping in other pages you've been updating, so I'm guessing this was just an oversight...)

laurapacilio replied:
oh yuck yes - thank you I'm going to fix this :)

apparentlymart commented on website/docs/cli/workspaces/index.mdx:
```diff
@@ -7,72 +7,99 @@ description: >-
+
+## Use Cases
+
+You can create multiple [working directories](/cli/init) to maintain multiple instances of a configuration with completely separate state data. However, Terraform installs a separate cache of plugins and modules for each working directory, so maintaining multiple directories can waste bandwidth and disk space. This approach also requires extra tasks like updating configuration from version contro ...
```
Comparing this to the old version of the page I see what you've done here is hoist the "what are workspaces useful for?" content up to the top, which seems reasonable given that workspaces are what this page is about.

I worry a little that this change in emphasis has subtly changed the subtext, though. The previous version of the page is essentially saying "Most of the time you don't need the extra complexity of workspaces, but there are a few benefits that _might_ apply to your situation." This version starts with "Here are all of the benefits of workspaces!", which makes it sound like we're recommending multiple workspaces as the default way to use Terraform.

I'm not sure what to suggest to resolve this. This new version is far more concise and I like it a lot, so I'm certainly not meaning to suggest we bring back all of the old waffly text.

What I'd like to avoid is readers of this page being left with the impression that workspaces are a crucial part of using Terraform CLI without Terraform Cloud, since I don't think that's really true... instead, they exist primarily to support Terraform Cloud and _may_ be useful in some certain situations without Terraform Cloud.

But for most people workspaces will probably just make Terraform harder to use, because they will now have one extra ambient setting to think about every time they are about to run `terraform apply` and will inevitably also need to start adding extra `-var` and `-var-file` options to provide the workspace-specific input variables. Without workspaces all of that information lives inside the configuration itself and so you can just run `terraform apply` and not worry about it. (For those who are using Terraform Cloud, the Terraform Cloud-specific features remove most of this complexity, and so multiple workspaces is far more useful for Terraform Cloud users.)

I'm sorry this isn't a very actionable comment. If you'd like to talk about this in a more synchronous way then I'd be happy to.

laurapacilio replied:
no no this is super helpful, Martin, and I think you're right. I really appreciate you calling this out. Let me think through how to fix this. i think I can rework a few strategic things to give users the right impression. You're right - right now it's like YAY WORKSPACES ARE SO. HELPFUL, when in reality they can be helpful in some circumstances and are available to you....but you really should only use them for specific things and they do also come with drawbacks.

laurapacilio commented:
Okay @apparentlymart thank you as usual for the excellent feedback on this :) I think addressed all of your comments. Specifically, I modified the introduction to:
- Immediately and directly contrast these with TFC workspaces, and provide a link to the TFC workspaces page so folks can learn more if they want to.
- Moved the thing about variables from the intro and put it in "Managing Workspaces" instead. Now, I phrased it more as like a "hey you need to remember to do this if you're deploying infrastructure with workspaces." Do you think it needs more, or okay as is?
- Took out (I think?) all of the remaining hard breaks in the text.
- Revised the introduction to....make it clearer that this isn't required and that it's really only useful in specific circumstances. 

Thoughts? I think this is better, but I am happy to address any further feedback you have.  Thank you again!!

apparentlymart reviewed (approved):
This looks great to me, and much better than the old content! Thanks for working on this.

laurapacilio commented on website/docs/language/state/workspaces.mdx:
````diff
@@ -7,71 +7,40 @@ description: >-
-so if you run "terraform plan" Terraform will not see any existing state
-for this configuration.
-```
+~> **Important:** Workspaces are not appropriate for system decomposition or deployments requiring separate credentials and access controls. Refer to [When Not to Use Multiple Workspaces](/cli/workspaces#when-not-to-use-multiple-workspaces) in the Terraform CLI documentation for details and recommended alternatives.
````
```suggestion
~> **Important:** Workspaces are not appropriate for system decomposition or deployments requiring separate credentials and access controls. Refer to [Use Cases](/cli/workspaces#use-cases) in the Terraform CLI documentation for details and recommended alternatives.
```

laurapacilio commented on website/docs/language/state/workspaces.mdx:
```diff
@@ -7,71 +7,40 @@ description: >-
-Each Terraform configuration has an associated [backend](/language/settings/backends)
-that defines how operations are executed and where persistent data such as
-[the Terraform state](/language/state/purpose) are
+Each Terraform configuration has an associated [backend](/language/settings/backends) that defines how operations are executed and where persistent data such as [the Terraform state](/language/state/purpose) are
```
```suggestion
Each Terraform configuration has an associated [backend](/language/settings/backends) that defines how Terraform executes operations where Terraform stores persistent data, like [Terraform state](/language/state/purpose).
```

laurapacilio commented on website/docs/language/state/workspaces.mdx:
```diff
@@ -7,71 +7,40 @@ description: >-
-that defines how operations are executed and where persistent data such as
-[the Terraform state](/language/state/purpose) are
+Each Terraform configuration has an associated [backend](/language/settings/backends) that defines how Terraform executes operations where Terraform stores persistent data, like [Terraform state](/language/state/purpose).
 stored.
```
```suggestion
```

laurapacilio commented on website/docs/language/state/workspaces.mdx:
```diff
@@ -7,71 +7,39 @@ description: >-
-to be associated with a single configuration. The configuration still
-has only one backend, but multiple distinct instances of that configuration
-to be deployed without configuring a new backend or changing authentication
+Each Terraform configuration has an associated [backend](/language/settings/backends) that defines how Terraform executes operations where Terraform stores persistent data, like [Terraform state](/language/state/purpose).
```
```suggestion
Each Terraform configuration has an associated [backend](/language/settings/backends) that defines how Terraform executes operations and where Terraform stores persistent data, like [state](/language/state/purpose).
```
